rust-spec/lightclient/verification/Blockchain_003_draft.tla

```diff
@@ -0,0 +1,164 @@
+
+(*
+ Given a function pVotingPower \in D -> Powers for some D \subseteq AllNodes
+ and pNodes \subseteq D, test whether the set pNodes \subseteq AllNodes has
+ more than 2/3 of voting power among the nodes in D.
+ *)
+TwoThirds(pVS, pNodes) ==
+    LET TP == Cardinality(pVS)
+        SP == Cardinality(pVS \intersect pNodes)
+    IN
+    3 * SP > 2 * TP \* when thinking in real numbers, not integers: SP > 2.0 / 3.0 * TP 
+
+(*
+ Given a set of FaultyNodes, test whether the voting power of the correct nodes in D
+ is more than 2/3 of the voting power of the faulty nodes in D.
+
+ Parameters:
+   - pFaultyNodes is a set of nodes that are considered faulty
+   - pVS is a set of all validators, maybe including Faulty, intersecting with it, etc.
+   - pMaxFaultRatio is a pair <<a, b>> that limits the ratio a / b of the faulty
+     validators from above (exclusive)
+ *)
+FaultyValidatorsFewerThan(pFaultyNodes, pVS, maxRatio) ==
+    LET FN == pFaultyNodes \intersect pVS   \* faulty nodes in pNodes
+        CN == pVS \ pFaultyNodes            \* correct nodes in pNodes
+        CP == Cardinality(CN)               \* power of the correct nodes
+        FP == Cardinality(FN)               \* power of the faulty nodes
+    IN
+    \* CP + FP = TP is the total voting power
+    LET TP == CP + FP IN
+    FP * maxRatio[2] < TP * maxRatio[1]
+
+(* Can a block be produced by a correct peer, or an authenticated Byzantine peer *)
+IsLightBlockAllowedByDigitalSignatures(ht, block) == 
+    \/ block.header = blockchain[ht] \* signed by correct and faulty (maybe)
+    \/ /\ block.Commits \subseteq Faulty
+       /\ block.header.height = ht
+       /\ block.header.time >= 0 \* signed only by faulty
+
+(*
+ Initialize the blockchain to the ultimate height right in the initial states.
+ We pick the faulty validators statically, but that should not affect the light client.
+
+ Parameters:
+    - pMaxFaultyRatioExclusive is a pair <<a, b>> that bound the number of
+        faulty validators in each block by the ratio a / b (exclusive)
+ *)            
+InitToHeight(pMaxFaultyRatioExclusive) ==
+  /\ Faulty \in SUBSET AllNodes \* some nodes may fail
+  \* pick the validator sets and last commits
+  /\ \E vs, lastCommit \in [Heights -> SUBSET AllNodes]:
+     \E timestamp \in [Heights -> Int]:
+        \* now is at least as early as the timestamp in the last block 
+        /\ \E tm \in Int: now = tm /\ tm >= timestamp[ULTIMATE_HEIGHT]
+        \* the genesis starts on day 1     
+        /\ timestamp[1] = 1
+        /\ vs[1] = AllNodes
+        /\ lastCommit[1] = EmptyNodeSet
+        /\ \A h \in Heights \ {1}:
+          /\ lastCommit[h] \subseteq vs[h - 1]   \* the non-validators cannot commit 
+          /\ TwoThirds(vs[h - 1], lastCommit[h]) \* the commit has >2/3 of validator votes
+            \* the faulty validators have the power below the threshold
+          /\ FaultyValidatorsFewerThan(Faulty, vs[h], pMaxFaultyRatioExclusive)
```

Comment:
Here we ascertain the faulty ratio `pMaxFaultyRatioExclusive` for all heights `h`. In the security model it would be 1/3. To prove the diamond property we have put 2/3 (there is less than 2/3 faulty voting power.